File: 14_并查集/经典题/每个对选一个点(每条边选一个点)/SelectOneFromEachPair.py
from collections import defaultdict
from typing import DefaultDict, List, Tuple


class SelectOneFromEachPairMap:
    """
    可撤销并查集，维护连通分量为树的联通分量个数.
    字典实现.
    """

    __slots__ = ("part", "treeCount", "_data", "_edge", "_history")

    def __init__(self):
        self.part = 0
        self.treeCount = 0  # 联通分量为树的联通分量个数(孤立点也算树)
        self._data = dict()
        self._edge = dict()
        self._history = []  # (root,data,edge)

    def solve(self) -> int:
        """
        从每条边中恰好选一个点, 最多能选出多少个不同的点.
        对每个大小为m的连通块,树的贡献为m-1,环的贡献为m.
        因此答案为`总点数-树的个数`.
        """
        return len(self._data) - self.treeCount

    def undo(self) -> bool:
        """撤销上一次合并操作，没合并成功也要撤销."""
        if not self._history:
            return False
        small, smallData, smallEdge = self._history.pop()
        big, bigData, bigEdge = self._history.pop()
        self._data[small] = smallData
        self._data[big] = bigData
        self._edge[small] = smallEdge
        self._edge[big] = bigEdge
        if big == small:
            if self.isTree(big):
                self.treeCount += 1
        else:
            if self.isTree(big) or self.isTree(small):
                self.treeCount += 1
            self.part += 1
        return True

    def isTree(self, u: int) -> bool:
        """u所在的联通分量是否为树."""
        root = self.find(u)
        vertex = self.getSize(root)
        return vertex == self._edge[root] + 1

    def countTree(self) -> int:
        """联通分量为树的联通分量个数(孤立点也算树)."""
        return self.treeCount

    def union(self, u: int, v: int) -> bool:
        """添加边对(u,v)."""
        u, v = self.find(u), self.find(v)
        self._history.append((u, self._data[u], self._edge[u]))  # big
        self._history.append((v, self._data[v], self._edge[v]))  # small
        if u == v:
            if self.isTree(u):
                self.treeCount -= 1
            self._edge[u] += 1
            return False
        if self._data[u] > self._data[v]:
            u, v = v, u
        if self.isTree(u) or self.isTree(v):
            self.treeCount -= 1
        self._data[u] += self._data[v]
        self._data[v] = u
        self._edge[u] += self._edge[v] + 1
        self.part -= 1
        return True

    def find(self, u: int) -> int:
        """不能路径压缩."""
        if u not in self._data:
            self.add(u)
            return u
        cur = u
        while self._data.get(cur, -1) >= 0:
            cur = self._data[cur]
        return cur

    def getSize(self, x: int) -> int:
        return -self._data[self.find(x)]

    def getEdge(self, x: int) -> int:
        return self._edge[self.find(x)]

    def getGroups(self) -> DefaultDict[int, List[int]]:
        groups = defaultdict(list)
        for k in self._data:
            root = self.find(k)
            groups[root].append(k)
        return groups

    def add(self, u: int) -> bool:
        """添加点u."""
        if u in self._data:
            return False
        self._data[u] = -1
        self._edge[u] = 0
        self.part += 1
        self.treeCount += 1
        return True

    def __repr__(self) -> str:
        return "\n".join(f"{root}: {member}" for root, member in self.getGroups().items())


class SelectOneFromEachPairArray:
    """
    可撤销并查集，维护连通分量为树的联通分量个数.
    数组实现.
    """

    __slots__ = ("part", "treeCount", "_n", "_data", "_edge", "_history")

    def __init__(self, n: int):
        self.part = n
        self.treeCount = n  # 联通分量为树的联通分量个数(孤立点也算树)
        self._n = n
        self._data = [-1] * n
        self._edge = [0] * n
        self._history = []  # (root,data,edge)

    def solve(self) -> int:
        """
        从每条边中恰好选一个点, 最多能选出多少个不同的点.
        对每个大小为m的连通块,树的贡献为m-1,环的贡献为m.
        因此答案为`总点数-树的个数`.
        """
        return self._n - self.treeCount

    def undo(self) -> bool:
        """撤销上一次合并操作，没合并成功也要撤销."""
        if not self._history:
            return False
        small, smallData, smallEdge = self._history.pop()
        big, bigData, bigEdge = self._history.pop()
        self._data[small] = smallData
        self._data[big] = bigData
        self._edge[small] = smallEdge
        self._edge[big] = bigEdge
        if big == small:
            if self.isTree(big):
                self.treeCount += 1
        else:
            if self.isTree(big) or self.isTree(small):
                self.treeCount += 1
            self.part += 1
        return True

    def isTree(self, u: int) -> bool:
        """u所在的联通分量是否为树."""
        root = self.find(u)
        vertex = self.getSize(root)
        return vertex == self._edge[root] + 1

    def countTree(self) -> int:
        """联通分量为树的联通分量个数(孤立点也算树)."""
        return self.treeCount

    def union(self, u: int, v: int) -> bool:
        """添加边对(u,v)."""
        u, v = self.find(u), self.find(v)
        self._history.append((u, self._data[u], self._edge[u]))  # big
        self._history.append((v, self._data[v], self._edge[v]))  # small
        if u == v:
            if self.isTree(u):
                self.treeCount -= 1
            self._edge[u] += 1
            return False
        if self._data[u] > self._data[v]:
            u, v = v, u
        if self.isTree(u) or self.isTree(v):
            self.treeCount -= 1
        self._data[u] += self._data[v]
        self._data[v] = u
        self._edge[u] += self._edge[v] + 1
        self.part -= 1
        return True

    def find(self, u: int) -> int:
        """不能路径压缩."""
        cur = u
        while self._data[cur] >= 0:
            cur = self._data[cur]
        return cur

    def getSize(self, x: int) -> int:
        return -self._data[self.find(x)]

    def getEdge(self, x: int) -> int:
        return self._edge[self.find(x)]

    def getGroups(self) -> DefaultDict[int, List[int]]:
        groups = defaultdict(list)
        for i in range(self._n):
            root = self.find(i)
            groups[root].append(i)
        return groups

    def __repr__(self) -> str:
        return "\n".join(f"{root}: {member}" for root, member in self.getGroups().items())


if __name__ == "__main__":
    # https://atcoder.jp/contests/arc111/tasks/arc111_b
    # B - Reversible Cards
    # !给定一些卡片,正反面标有数字,你需要反转一些卡片
    # !使得所有卡片的正面数字种类最多

    # 正反面连边,问题变为选取每个边的一个端点，问最多选取多少个
    # 对每个大小为n的连通块
    # 树: n-1
    # 有环: n

    def reversibleCards(cards: List[Tuple[int, int]]) -> int:
        uf = SelectOneFromEachPairMap()
        for u, v in cards:
            uf.union(u, v)
        return uf.solve()

    # https://atcoder.jp/contests/abc302/tasks/abc302_h
    # 给定一棵树，每个点有两个值。
    # 对于v=1,2,3,...,n，问从点1到点 v的最短路径途径的每个点中，
    # 各选一个数，其不同数的个数的最大值。
    def ballCollector(
        n: int, edges: List[Tuple[int, int]], pairs: List[Tuple[int, int]]
    ) -> List[int]:
        def dfs(cur: int, pre: int) -> None:
            a, b = pairs[cur]
            uf.union(a, b)
            res[cur] = uf.solve()
            for next in adjList[cur]:
                if next != pre:
                    dfs(next, cur)
            uf.undo()

        uf = SelectOneFromEachPairMap()
        adjList = [[] for _ in range(n)]
        for u, v in edges:
            adjList[u].append(v)
            adjList[v].append(u)
        res = [0] * n
        dfs(0, -1)
        return res

    import sys

    sys.setrecursionlimit(int(1e9))
    input = lambda: sys.stdin.readline().rstrip("\r\n")

    n = int(input())
    pairs = [tuple(map(int, input().split())) for _ in range(n)]
    edges = []
    for _ in range(n - 1):
        u, v = map(int, input().split())
        edges.append((u - 1, v - 1))
    res = ballCollector(n, edges, pairs)
    print(*res[1:])

File: 14_并查集/经典题/每个对选一个点(每条边选一个点)/SelectOneFromEachPair.ts
/* eslint-disable @typescript-eslint/no-non-null-assertion */

/**
 * 可撤销并查集，维护连通分量为树的联通分量个数.
 * Map 实现.
 */
class SelectOneFromEachPairMap {
  private _part = 0
  private _treeCount = 0
  private readonly _data = new Map<number, number>()
  private readonly _edge = new Map<number, number>()
  private readonly _history: [root: number, data: number, edge: number][] = []

  /**
   * 从每条边中恰好选一个点, 最多能选出多少个不同的点.
   * 对每个大小为m的连通块,树的贡献为m-1,环的贡献为m.
   * 因此答案为`总点数-树的个数`.
   */
  solve(): number {
    return this._data.size - this._treeCount
  }

  /**
   * 撤销上一次合并操作，没合并成功也要撤销.
   */
  undo(): boolean {
    if (!this._history.length) {
      return false
    }
    const [small, smallData, smallEdge] = this._history.pop()!
    const [big, bigData, bigEdge] = this._history.pop()!
    this._data.set(small, smallData)
    this._data.set(big, bigData)
    this._edge.set(small, smallEdge)
    this._edge.set(big, bigEdge)
    if (big === small) {
      if (this.isTree(big)) {
        this._treeCount++
      }
    } else {
      if (this.isTree(big) || this.isTree(small)) {
        this._treeCount++
      }
      this._part++
    }
    return true
  }

  /**
   * u所在的联通分量是否为树.
   */
  isTree(u: number): boolean {
    const root = this.find(u)
    const vertex = this.getSize(root)
    return vertex === this._edge.get(root)! + 1
  }

  /**
   * 联通分量为树的联通分量个数(孤立点也算树).
   */
  countTree(): number {
    return this._treeCount
  }

  /**
   * 添加边对(u,v).
   * @alias addPair
   */
  union(u: number, v: number): boolean {
    u = this.find(u)
    v = this.find(v)
    this._history.push([u, this._data.get(u)!, this._edge.get(u)!]) // big
    this._history.push([v, this._data.get(v)!, this._edge.get(v)!]) // small
    if (u === v) {
      if (this.isTree(u)) {
        this._treeCount--
      }
      this._edge.set(u, this._edge.get(u)! + 1)
      return false
    }

    if (this._data.get(u)! > this._data.get(v)!) {
      u ^= v
      v ^= u
      u ^= v
    }
    if (this.isTree(u) || this.isTree(v)) {
      this._treeCount--
    }
    this._data.set(u, this._data.get(u)! + this._data.get(v)!)
    this._data.set(v, u)
    this._edge.set(u, this._edge.get(u)! + this._edge.get(v)! + 1)
    this._part--
    return true
  }

  /**
   * 不能路径压缩.
   */
  find(u: number): number {
    if (!this._data.has(u)) {
      this.add(u)
      return u
    }
    let cur = u
    while ((this._data.get(cur) || -1) >= 0) {
      cur = this._data.get(cur)!
    }
    return cur
  }

  getSize(x: number): number {
    return -this._data.get(this.find(x))!
  }

  getEdge(x: number): number {
    return this._edge.get(this.find(x))!
  }

  getGroups(): Map<number, number[]> {
    const groups = new Map<number, number[]>()
    for (const k of this._data.keys()) {
      const root = this.find(k)
      if (!groups.has(root)) {
        groups.set(root, [])
      }
      groups.get(root)!.push(k)
    }
    return groups
  }

  add(u: number): boolean {
    if (this._data.has(u)) {
      return false
    }
    this._data.set(u, -1)
    this._edge.set(u, 0)
    this._part++
    this._treeCount++
    return true
  }

  get part(): number {
    return this._part
  }

  get treeCount(): number {
    return this._treeCount
  }
}

/**
 * 可撤销并查集，维护连通分量为树的联通分量个数.
 * 数组实现.
 */
class SelectOneFromEachPairArray {
  private _part: number
  private _treeCount: number
  private readonly _n: number
  private readonly _data: Int32Array
  private readonly _edge: Uint32Array
  private readonly _history: [root: number, data: number, edge: number][] = []

  constructor(n: number) {
    this._part = n
    this._treeCount = n
    this._n = n
    this._data = new Int32Array(n).fill(-1)
    this._edge = new Uint32Array(n)
  }

  /**
   * 从每条边中恰好选一个点, 最多能选出多少个不同的点.
   * 对每个大小为m的连通块,树的贡献为m-1,环的贡献为m.
   * 因此答案为`总点数-树的个数`.
   */
  solve(): number {
    return this._n - this._treeCount
  }

  /**
   * 撤销上一次合并操作，没合并成功也要撤销.
   */
  undo(): boolean {
    if (!this._history.length) {
      return false
    }
    const [small, smallData, smallEdge] = this._history.pop()!
    const [big, bigData, bigEdge] = this._history.pop()!
    this._data[small] = smallData
    this._data[big] = bigData
    this._edge[small] = smallEdge
    this._edge[big] = bigEdge
    if (big === small) {
      if (this.isTree(big)) {
        this._treeCount++
      }
    } else {
      if (this.isTree(big) || this.isTree(small)) {
        this._treeCount++
      }
      this._part++
    }
    return true
  }

  /**
   * u所在的联通分量是否为树.
   */
  isTree(u: number): boolean {
    const root = this.find(u)
    const vertex = this.getSize(root)
    return vertex === this._edge[root] + 1
  }

  /**
   * 联通分量为树的联通分量个数(孤立点也算树).
   */
  countTree(): number {
    return this._treeCount
  }

  /**
   * 添加边对(u,v).
   * @alias addPair
   */
  union(u: number, v: number): boolean {
    u = this.find(u)
    v = this.find(v)
    this._history.push([u, this._data[u], this._edge[u]]) // big
    this._history.push([v, this._data[v], this._edge[v]]) // small
    if (u === v) {
      if (this.isTree(u)) {
        this._treeCount--
      }
      this._edge[u]++
      return false
    }

    if (this._data[u] > this._data[v]) {
      u ^= v
      v ^= u
      u ^= v
    }
    if (this.isTree(u) || this.isTree(v)) {
      this._treeCount--
    }
    this._data[u] += this._data[v]
    this._data[v] = u
    this._edge[u] += this._edge[v] + 1
    this._part--
    return true
  }

  /**
   * 不能路径压缩.
   */
  find(u: number): number {
    let cur = u
    while (this._data[cur] >= 0) {
      cur = this._data[cur]
    }
    return cur
  }

  getSize(x: number): number {
    return -this._data[this.find(x)]
  }

  getEdge(x: number): number {
    return this._edge[this.find(x)]
  }

  getGroups(): Map<number, number[]> {
    const groups = new Map<number, number[]>()
    for (let i = 0; i < this._n; i++) {
      const root = this.find(i)
      if (!groups.has(root)) {
        groups.set(root, [])
      }
      groups.get(root)!.push(i)
    }
    return groups
  }

  get part(): number {
    return this._part
  }

  get treeCount(): number {
    return this._treeCount
  }
}

export { SelectOneFromEachPairMap }

/**
 * 给定一棵树，每个点有两个值。
 * 对于v=1,2,3,...,n，问从点1到点 v的最短路径途径的每个点中，
 * 各选一个数，其不同数的个数的最大值。
 * @link https://atcoder.jp/contests/abc302/tasks/abc302_h
 */
function ballCollector(n: number, edges: [number, number][], pairs: [number, number][]): number[] {
  const uf = new SelectOneFromEachPairMap()
  const adjList: number[][] = Array(n)
  for (let i = 0; i < n; i++) adjList[i] = []
  edges.forEach(([u, v]) => {
    adjList[u].push(v)
    adjList[v].push(u)
  })

  const res: number[] = Array(n).fill(0)
  dfs(0, -1)
  return res

  function dfs(cur: number, pre: number): void {
    const [a, b] = pairs[cur]
    uf.union(a, b)
    res[cur] = uf.solve()
    adjList[cur].forEach(next => {
      if (next !== pre) {
        dfs(next, cur)
      }
    })
    uf.undo()
  }
}
